### PR TITLE
bugfix: Fixes #3100 rts_memory_size overflow; add non-ci test exercising overflow detection in rts alloc_words

### LINKED ISSUE #3100: bug: `rts_memory_size` overflows when reporting last page
`rts_memory_size` incorrectly reports 0 (bytes) when wasm `memory.size` is 65536 (pages)

Due to multiplication by `page_size` in 32bit, before extending to `Nat`.

### PR DESCRIPTION
**Pull request by crusso**
* Fix for #3100 
* add preupgrade-oom.mo test to exercise overflow detection in pre-upgrade
  (should fail silently on #3095) and < 0.6.21 versions of moc.

### CODE REVIEW

File: test/run-drun/upgrade-bug-immut.mo [hunk 1]
```diff
@@ -1,11 +1,16 @@
 import P "mo:⛔";
 
 actor {
-  stable var a : [Nat] = P.Array_tabulate<Nat>(268435456 / 4, func _ { 0x0F } ); // 0.25 GB array (I think)
+  stable var a : [Nat] = [];
 
-  system func preupgrade() { P.debugPrint("pre"); };
+  system func preupgrade() {
+   a := P.Array_tabulate<Nat>(268435456 / 4, func _ { 0x0F } ); // 0.25 GB array (I think)
+   P.debugPrint("pre"); };
```
Comment (crusso) [inline test/run-drun/upgrade-bug-immut.mo:8]:
```suggestion
   P.debugPrint("pre"); 
   };
```

**Comment by crusso** [comment]
@ggreif @nomeata any chance on a quick review of this - it fixes an overflow and adds a low memory test.

File: src/codegen/compile.ml [hunk 2]
```diff
@@ -904,7 +904,8 @@ module Heap = struct
 
   let get_memory_size =
     G.i MemorySize ^^
-    compile_mul_const page_size
+    G.i (Convert (Wasm.Values.I64 I64Op.ExtendUI32)) ^^
```
Comment (ggreif) [inline src/codegen/compile.ml:907]: Good catch!

**Review by ggreif — APPROVED** [review]
LGTM! Since `pre` traps, do you need `post` at all?
